File: components/redis-handling-session/src/main/java/com/sv/youapp/component/authorization/entities/redis/OAuth2AuthorizationCodeGrantAuthorization.java
package com.sv.youapp.component.authorization.entities.redis;

import lombok.Getter;
import org.springframework.data.redis.core.index.Indexed;
import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationRequest;

import java.security.Principal;
import java.time.Instant;

@Getter
public class OAuth2AuthorizationCodeGrantAuthorization extends AbstractGrantAuthorization {

	private final Principal principal;
	private final OAuth2AuthorizationRequest authorizationRequest;
	private final AuthorizationCode authorizationCode;

	@Indexed
	private final String state;

	public OAuth2AuthorizationCodeGrantAuthorization(
		String id,
		String registeredClientId,
		String principalName,
		String scopes,
		AccessToken accessToken,
		RefreshToken refreshToken,
		Principal principal,
		OAuth2AuthorizationRequest authorizationRequest,
		AuthorizationCode authorizationCode,
		String state
	) {
		super(id, registeredClientId, principalName, scopes, accessToken, refreshToken);
		this.principal = principal;
		this.authorizationRequest = authorizationRequest;
		this.authorizationCode = authorizationCode;
		this.state = state;
	}

	public static final class AuthorizationCode extends AbstractToken {
		public AuthorizationCode(
			String tokenValue,
			Instant issuedAt,
			Instant expiresAt,
			boolean invalidated
		) {
			super(tokenValue, issuedAt, expiresAt, invalidated);
		}
	}
}

File: authorization-server/src/main/kotlin/com/sv/youapp/authorization/entities/redis/OAuth2AuthorizationCodeGrantAuthorization.kt
package com.sv.youapp.authorization.entities.redis

import org.springframework.data.redis.core.index.Indexed
import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationRequest
import java.security.Principal
import java.time.Instant

open class OAuth2AuthorizationCodeGrantAuthorization(
    id: String,
    registeredClientId: String,
    principalName: String,
    scopes: String,
    accessToken: AccessToken?,
    refreshToken: RefreshToken?,
    val principal: Principal?,
    val authorizationRequest: OAuth2AuthorizationRequest?,
    val authorizationCode: AuthorizationCode?,
    @Indexed
    val state: String?,
) : AbstractGrantAuthorization(id, registeredClientId, principalName, scopes, accessToken, refreshToken) {

    class AuthorizationCode(tokenValue: String?, issuedAt: Instant?, expiresAt: Instant?, invalidated: Boolean) :
        AbstractToken(tokenValue, issuedAt, expiresAt, invalidated)
}
